raft: stop group0 server during group0 service shutdown

margdoc opened the pull request:
When the `topology_change` command is applied, the topology state is reloaded and `cdc::generation_service::handle_cdc_generation` is called. This creates a dependency of the group0 on `cdc::generation_service`.
Currently, the group0 server is stopped during `raft_group_registry` shutdown. However, it is called after `cdc::generation_service` shutdown, which can result in a segfault.
  
To prevent this issue, this commit stops the group0 server and removes it from the `raft_group_registry` during the `group0_service` shutdown.

Fixes #14397.

Reproducer: https://github.com/margdoc/scylla/commit/97d6946e314a5b5d445b57045d93c57d10494b48.
It creates two nodes. The second one is forced to stop after joining the group0. It sleeps before calling `handle_cdc_generation` and sleeps just before `raft_group_registry` is stopped. It ensures that `handle_cdc_generation` wakes up after starting the second sleep. If the `cdc_generation_service` shutdown waits for `raft_group_registry` to stop, `handle_cdc_generation`  will be called without any issue. Otherwise, it will crash since `cdc_generation_service` won't exist. The test passes always. If the crash happens it can be seen in the log file of the second node.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2622/

kbr-scylla commented:
> This creates a dependency of cdc::generation_service on the group0.

Don't you mean the other direction -- group 0 depends on generation_service?

kbr-scylla commented on main.cc:
```diff
@@ -1520,6 +1513,12 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
+            // XXX: stop_raft has to happen before query_processor, migration_manager and cdc_generation_service
+            // are stopped, since some groups keep using these services until are stopped inside stop_raft.
+            auto stop_raft = defer_verbose_shutdown("Raft", [&raft_gr] {
```
This approach is wrong. We can't just keep moving raft_gr stopping later and later because "some groups" depend on some services.

We must stop the groups which have the dependency.

So the order of init/stop should be:

- start raft_group_registry
- start service X
- start a group which depends on X
- ... running ...
- stop the group which depends on X
- stop service X
- stop raft group registry.

OR:
- start services X, Y, ...
- start raft group registry and all groups which depend on X, Y, ...
- ... running ...
- stop raft group registry and all the groups which depend on X, Y...
- stop services X, Y, ...

In general, stopping order should be the reverse of starting order.

cc @kostja 

I see it was broken already before:
```cpp
            // XXX: stop_raft has to happen before query_processor and migration_manager
            // is stopped, since some groups keep using the query
            // processor until are stopped inside stop_raft.
            auto stop_raft = defer_verbose_shutdown("Raft", [&raft_gr] {
                raft_gr.stop().get();
            });
```

maybe this is a good opportunity to fix it.

kbr-scylla replied:
This is where group 0 is being setup:
```cpp

            // Set up group0 service earlier since it is needed by group0 setup just below
            ss.local().set_group0(group0_service);

            // Setup group0 early in case the node is bootstrapped already and the group exists.
            // Need to do it before allowing incomming messaging service connections since
            // storage proxy's and migration manager's verbs may access group0.
            // This will also disable migration manager schema pulls if needed.
            group0_service.setup_group0_if_exist(sys_ks.local(), ss.local(), qp.local(), mm.local(), cdc_generation_service.local()).get();

            // It's essential to load fencing_version prior to starting the messaging service,
            // since incoming messages may require fencing.
            // It is also necessary that token_metadata be initialized by this point and
            // token_metadata::version to be up to date, since when assigning fence_version we
            // validate it cannot be greater than token_metadata::version.
            // The setup_group0_if_exist call above waits for the raft log for group0 to be applied,
            // which ensures this.
            ss.local().update_fence_version(sys_ks.local().get_topology_fence_version().get()).get();

            with_scheduling_group(maintenance_scheduling_group, [&] {
                return messaging.invoke_on_all(&netw::messaging_service::start_listen, std::ref(token_metadata));
            }).get();

            with_scheduling_group(maintenance_scheduling_group, [&] {
                return ss.local().join_cluster(cdc_generation_service.local(), sys_dist_ks, proxy, qp.local());
            }).get();
```

Therefore this is where the dependency on CDC gen service and other services is introduced during startup.
Therefore, this is also where we should have a `stop_group0` deferred callback.

kbr-scylla replied:
This is also sloppy (the comment):
```cpp
            // group0 client exists only on shard 0.
            // The client has to be created before `stop_raft` since during
            // destruction it has to exist until raft_gr.stop() completes.
            service::raft_group0_client group0_client{raft_gr.local(), sys_ks.local()};
```

the comment argues that we need to stop_raft() before destroying group0_client.
But again, this is only because of the broken deinit which is missing a `stop_group0_client` step.

We must start fixing our service start/stop order, which should now be easy thanks to the sp/qp/sys_ks refactors. We should not be introducing more hacks into it or extending existing hacks.

cc @gleb-cloudius

kbr-scylla replied:
Every service construction+start should have a corresponding stop+destruct in a deferred callback that is deferred right after we start the service.

Similarly, if a cyclic dependency is introduced between services after they are started, there should be a deferred callback (right below introducing the dependency) that breaks the dependency.

So the code should look like:
```
X.start();
defer(stop_X);

Y.start(ref(X));
defer(stop_Y);

Z.start(ref(Y));
defer(stop_Z);

add_dependency_y_to_z(ref(Y), ref(Z));
defer(remove_dependency_y_to_z);
```

and so on.

cc @xemul

gleb-cloudius replied:
> This is also sloppy (the comment):
> 
> ```c++
>             // group0 client exists only on shard 0.
>             // The client has to be created before `stop_raft` since during
>             // destruction it has to exist until raft_gr.stop() completes.
>             service::raft_group0_client group0_client{raft_gr.local(), sys_ks.local()};
> ```
> 
> the comment argues that we need to stop_raft() before destroying group0_client. But again, this is only because of the broken deinit which is missing a `stop_group0_client` step.

In general I agree that stopping all groups together  will not fly in the long run, but since the only group we have right now is group0 what you propose will simply rename things and will not change anything substantial. So here the comment will say that "The client has to be created before `stop_group0` since during destruction it has to exist until group0.stop() completes". Yes it is more correct. but today stopping group0 and stopping raft_gr is one and the same.

If you think this is something that should be done right now, I will not stop you.

margdoc commented:
v2:
* changed the solution to stopping group0 server during group0 service shutdown

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2660/

kbr-scylla commented on service/raft/raft_group_registry.cc:
```diff
@@ -335,6 +335,18 @@ static void ensure_aborted(raft_server_for_group& server_for_group, sstring reas
+    }
+
+    ensure_aborted(it->second, std::move(reason));
+    co_await std::move(*it->second.aborted);
```
I think this could be called concurrently with `abort_server` which is called by `raft::server::configuration::on_background_error` for the group 0 server.

So the following could happen:
- here we move from the `aborted` future which causes it to turn into a ready future (I think? If not, and the behavior is undefined, it's even worse)
- while we're waiting, before we `erase`, concurrently `on_background_error` calls `abort_server` which sets `aborted` again
- then we `erase` the server, so this new `aborted` future is never waited for (it's discarded).

I think to prevent this, we should remove the server from the map before waiting for the future. Then `abort_server` won't find it.
But we cannot just `erase` it because that would destroy the server.

We should be able to move the server first, then erase, then wait for the abort to finish. Something like:
```cpp
auto srv = std::move(it->second);
_servers.erase(it);
co_await std::move(*srv.aborted);
```

kbr-scylla commented on service/raft/raft_group_registry.hh:
```diff
@@ -111,6 +111,9 @@ public:
     seastar::future<> stop();
 
+    // Stop the server for the given group and remove it from the registry.
+    seastar::future<> stop_server(raft::group_id gid, sstring reason);
```
Please mention the difference from `abort_server`.

kbr-scylla commented on service/raft/raft_group0.cc:
```diff
@@ -366,6 +366,8 @@ future<group0_info> persistent_discovery::run(
 future<> raft_group0::abort() {
+    co_await stop_group0();
+
     co_await smp::invoke_on_all([this]() {
```
RPC verbs are initialized first, *then* group 0 server is started.

The shutdown order should be the reverse of the init order.

In this case, imagine one of the RPC verbs accesses the group 0 server. If we stop and remove the server first, a verb handler may fire which depends on the server, trying to access a removed server.

In fact `stop_group0` should be the last step of `abort` -- after `_shutdown_gate.close()` -- the gate is held by tasks that are accessing the server too. So we should first close the gate, ensuring that none of these tasks are running, and only then stop the server.

margdoc commented:
v3:
* stop_group0: made it private and added a comment
* stop_server: mentioned the difference from abort_server
* moved stop_group0 call to the end of raft_group0::abort
* raft_group_registry::stop_server: erase from the map and then abort it

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2740/
